openvdb_ax/openvdb_ax/compiler/LeafLocalData.h

```diff
@@ -0,0 +1,229 @@
+#define OPENVDB_AX_COMPILER_LEAF_LOCAL_DATA_HAS_BEEN_INCLUDED
+
+#include <openvdb/openvdb.h>
+#include <openvdb/points/AttributeArray.h>
+#include <openvdb/points/PointAttribute.h>
+#include <openvdb/points/PointDataGrid.h>
+#include <openvdb/points/PointGroup.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+
+namespace ax {
+namespace compiler {
+
+
+/// @brief  Various functions can request the use and initialization of point data from within
+///         the kernel that does not use the standard attribute handle methods. This data can
+///         then be accessed after execution to perform post-processes such as adding new groups,
+///         adding new string attributes or updating positions.
+///
+/// @note  Due to the way string handles work, string write attribute handles cannot
+///        be constructed in parallel, nor can read handles retrieve values in parallel
+///        if there is a chance the shared metadata is being written to (with set()).
+///        As the compiler allows for any arbitrary string setting/getting, leaf local
+///        maps are used for temporary storage per point. The maps use the string array
+///        pointers as a key for later synchronization.
+///
+struct LeafLocalData
+{
+    using UniquePtr = std::unique_ptr<LeafLocalData>;
+    using GroupArrayT = openvdb::points::GroupAttributeArray;
+    using GroupHandleT = openvdb::points::GroupWriteHandle;
+
+    using PointStringMap = std::map<uint64_t, std::string>;
+    using StringArrayMap = std::map<points::AttributeArray*, PointStringMap>;
+
+    using LeafNode = openvdb::points::PointDataTree::LeafNodeType;
+
+    /// @brief  Construct a new data object to keep track of various data objects
+    ///         created per leaf by the point compute generator.
+    ///
+    /// @param  count  The number of points within the current leaf, used to initialize
+    ///                the size of new arrays
+    ///
+    LeafLocalData(const size_t count)
+        : mPointCount(count)
+        , mArrays()
+        , mOffset(0)
+        , mHandles()
+        , mStringMap() {}
+
+    ////////////////////////////////////////////////////////////////////////
+
+    /// Group methods
+
+    /// @brief  Return a group write handle to a specific group name, creating the
+    ///         group array if it doesn't exist. This includes either registering a
+    ///         new offset or allocating an entire array. The returned handle is
+    ///         guaranteed to be valid.
+    ///
+    /// @param  name  The group name
+    ///
+    inline GroupHandleT* getOrInsert(const std::string& name)
+    {
+        GroupHandleT* ptr = get(name);
+        if (ptr) return ptr;
+
+        static const size_t maxGroupsInArray =
+#if (OPENVDB_LIBRARY_MAJOR_VERSION_NUMBER > 7 ||  \
+    (OPENVDB_LIBRARY_MAJOR_VERSION_NUMBER >= 7 && \
+     OPENVDB_LIBRARY_MINOR_VERSION_NUMBER >= 1))
```

Comment:
A good reason to synchronize the versions IMO. How far back should we support this branching logic and when do we remove it?

Reply:
With the port into VDB, like with the deprecated code branching, I'll be removing all of this